Fix mongodb lookup

marcosdiez opened the pull request:
This PR has moved to https://github.com/ansible-collections/mongodb/pull/38
--------------------

##### SUMMARY
I am the original author of the plugin. 
After receiving an email with a support question, I realized:
  * the documentation has been update (by somebody else) and had an invalid ansible syntax, therefore it was not working
  * the plugin was not loudly complaining in case pymongo was not present
  * the plugin was not properly displaying error messages in ansible 2.8 ( when I wrote it, ansible was at version 2.3)

##### ISSUE TYPE
- Bugfix Pull Request
- Docs Pull Request

##### COMPONENT NAME

mongodb_lookup

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```
ansible 2.8.0.dev0 (fix_mongodb_lookup 8f744587e9) last updated 2018/10/06 13:52:09 (GMT -200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/mdiez/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /mnt/extra/u/rootfs/home/ubuntu/code/ansible/lib/ansible
  executable location = /mnt/extra/u/rootfs/home/ubuntu/code/ansible/bin/ansible
  python version = 3.5.2 (default, Nov 23 2017, 16:37:01) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

The biggest problem that I fixed was in the example usage in the documentation:

As one can see here: https://docs.ansible.com/ansible/latest/plugins/lookup/mongodb.html

it eventually says
```
    - debug: msg="Mongo has already started with the following PID [{{ item.pid }}]"
      with_mongodb: "{{mongodb_parameters}}"

```

which makes no sense since the mongodb lookup does not return an `item` object.

The second change is that sometime between `ansible 2.3` and `devel`, my plugin stopped writing error messages. So I am now using `AnsibleError`, which works !

Last but not least somehow I forgot to check if `pymongo` was installed. And it took me 3 years to find it out. Oops...

I also added integration tests to make sure things won't break in the future!

marcosdiez committed:
mongodb_lookup: fixed documentation and better error messages

skipping python2.6 on tests

ERROR: lib/ansible/plugins/lookup/mongodb.py:180:39: ansible-format-automatic-specification Format string contains automatic field numbering specification (100%)

one million of examples!

remove tab found by flake8

blank line after import

integration tests for lookup_mongodb

lint

cosmetic changes

we now have connection timeout in our tests

better documentation and tests on multiple ubutus

documentation changes

marcosdiez commented:
@Lujeni @Dufgui hi guys! Two years ago you reviewed my ansible mongodb lookup module. It has some issues and here is a PR with some code fixes. Would you care to code review them ?

Thanks!

Dufgui reviewed (approved):
Ok for me.

Lujeni commented on lib/ansible/plugins/lookup/mongodb.py:
```diff
@@ -18,6 +18,7 @@
 
 from __future__ import (absolute_import, division, print_function)
 from ansible.module_utils.six import string_types, integer_types
+from ansible.module_utils._text import to_native
```
not a big deal, you can add a space :)

marcosdiez replied:
I did not understand, sorry. Could you please explain it with other words ?

Lujeni replied:
blank line after your import

dagwieers replied:
These imports should go after the DOCUMENTATION, EXAMPLES and RETURN strings.

Lujeni commented on lib/ansible/plugins/lookup/mongodb.py:
```diff
@@ -81,20 +82,24 @@
-    - debug: msg="Mongo has already started with the following PID [{{ item.pid }}]"
-      with_mongodb: "{{mongodb_parameters}}"
+    - debug: msg="Mongo has already started with the following PID [{{ item.pid }}] - full_data {{ item }} "
+      with_items:
```
you should use `loop`, `with_items` is deprecated :)

Lujeni commented on lib/ansible/plugins/lookup/mongodb.py:
```diff
@@ -163,49 +164,20 @@ def convert_mongo_result_to_valid_json(self, result):
+
+    def _run_helper(self, terms):
+        if not pymongo_found:
+            raise AnsibleError(u"pymongo is required in the master node (this machine) for mongodb lookup.")
```
imho, you can keep your message KISS and more straighforward: 
e.g. "pymongo library is required for this module"

marcosdiez replied:
I would prefer not to, because this is a lookup module and therefore it must be in master node instead of the target node. Your message may confuse users who are not experts in ansible.

Lujeni replied:
in this case, you should use `primary` instead of `master`

marcosdiez replied:
understood.! fixed!

gundalow replied:
`control node` = Where ansible-playbook is run
`remote node` = system to configure

Lujeni reviewed (changes requested):
Thanks for the PR.
btw Can you rebase / squash your commit pls ?

Ciao

marcosdiez commented:
Hello @Lujeni !

I did some of the modifications you requested, added 5 more examples and some argumentation about other documentation. 
Could you please check now ?

mattclay commented:
bot_status

marcosdiez commented:
@Lujeni please please please please don't forget me !

marcosdiez commented:
@Lujeni please please please please please please please please  don't forget me !

marcosdiez commented:
@Lujeni @Lujeni @Lujeni @Lujeni @Lujeni @Lujeni  don't forget me !

marcosdiez commented:
@Lujeni  I even added integration tests.... all I need is an OK from you. Please...

marcosdiez commented:
@Lujeni wheeeeeeeeeeeeeeeeere aaaaaaaaaaaaaareeeeeeeeeee youuuuuuuuuuuuuuuu?

marcosdiez commented:
@Lujeni please code review my PR

mdiez-modus commented:
@Lujeni have you forgotten me?

marcosdiez commented:
@beak42py could you please code review ?

marcosdiez commented:
bot_status

marcosdiez commented:
@dagwieers is there anything I can do to help the PR get merged?

dagwieers commented:
We can start a MongoDB Working Group to get mongodb developers review and approve PRs. This could help moving things forward. I already tagged all MongoDB related issues and PRs, so it would only be a small step

marcosdiez commented:
@dagwieers it has been another 11 days and... nothing ? Can I help you getting this merged ? I have users who are counting on this bugfix

dagwieers commented:
@marcosdiez I can review, but I cannot test. And since I have no mongodb background I can't evaluate properly. So it is better to set up a group of peers to collaborate. Let me set up the MongoDB Working Group and then we can ask the group.

dagwieers commented on lib/ansible/plugins/lookup/mongodb.py:
```diff
@@ -62,7 +64,7 @@
         sort:
             description:
-                - Sorting rules. Please notice the constats are replaced by strings.
+                - Sorting rules. Please use the strings "ASCENDING" and "DESCENDING" to set the order. Check the example for more information.
```
Please split up the description items. The first item should be an explanation of the description.
```suggestion
                - Sorting rules.
                - Please use the strings C(ASCENDING) and C(DESCENDING) to set the order.information.
```

dagwieers reviewed (dismissed):
Only a few remarks. I cannot test, but the integration tests are promising.

dagwieers commented:
@marcosdiez 2 approvals have no meaning to @ansibot. What you need are 2 community shipits from maintainers. And that is not possible currently. Only if we have a MongoDB Working Group we can give members shipit rights.

gundalow commented:
@marcosdiez Looks like their are just a few minor things to address, could you take a look?

We are very close to the Ansible 2.8 cut off https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_8.rst

marcosdiez commented:
Hi John,

I'll check it tonight after my working hours.

Thanks for your time!

Marcos

On 2019-03-14 17:06, John R Barker wrote:
>
> @marcosdiez <https://github.com/marcosdiez> Looks like their are just 
> a few minor things to address, could you take a look?
>
> We are very close to the Ansible 2.8 cut off 
> https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_8.rst
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/ansible/ansible/pull/46579#issuecomment-473038827>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AASKGmwCtVL8tHU5Hsb7f-6udkZOOmDmks5vWqu_gaJpZM4XLee9>.
>

mdiez-modus commented:
@gundalow changes made :)

sivel commented:
@marcosdiez @mdiez-modus I don't see that any new commits have been pushed. I see you have commented from an account different than the one associated with this PR.  Did you push to the correct remote?

mdiez-modus commented:
@sivel I am not a very bright man... please check again :)

mdiez-modus commented:
All done! Rebased and all tests passed!

marcosdiez commented:
@dagwieers I made all the changes your requested.
@gundalow @sivel what else is need to have this merged ? The PR was actually done in October and it's a bugfix to something that I wrote myself. Should it really take that long?

mattclay commented on test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:
```diff
@@ -0,0 +1,62 @@
+- name: "assert we complain properly if we can't connect to mongo"
+  debug: msg="{{ lookup('mongodb', mongodb_parameters ) }}"
+  register: result
+  ignore_errors: yes
```
Can the connection timeout be lowered? This task takes 30 seconds to complete:

```
13:42 TASK [lookup_mongodb : assert we complain properly if we can't connect to mongo] ***
14:12 There was an exception on the mongodb_lookup: unable to connect to database: localhost:27017: [Errno 111] Connection refused
```

marcosdiez replied:
Agreed. I changed it to 5 seconds in the test.

mattclay commented on test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:
```diff
@@ -0,0 +1,62 @@
+- name: "assert we complain properly if pymongo is not there"
+  debug: msg="{{ lookup('mongodb', mongodb_parameters ) }}"
+  register: result
+  ignore_errors: yes
```
The test should not assume that `pymongo` is not already installed.

gundalow replied:
This doesn't look like it's been resolved

marcosdiez replied:
I believe it is. I check if "pymongo is required" exactly 5 lines below this one.

gundalow replied:
I think the problem is that `pymongo` may already be installed

mattclay commented on test/integration/targets/lookup_mongodb/tasks/test_lookup_mongo.yml:
```diff
@@ -0,0 +1,62 @@
+
+- name: Installing mongodb
+  # somehow the apt module messes up all tests...
+  command: "apt-get install -o Dpkg::Options::=\"--force-confold\" --force-yes -y mongodb"
```
What issue(s) did you encounter when trying to use the `apt` module?

gundalow replied:
@marcosdiez Could we please get an answer to this?

marcosdiez replied:
That's a hard one. I don't remember. I wrote this code in October. If this is blocking the PR from being merged, I can try to reproduce the issue. In my humble opinion this is secondary, since 
- this is only a test and not the main code itself
- yes, it could be better but it actually works

gundalow replied:
Agreed, I don't think this blocks

marcosdiez commented:
Hey, what else needs to be done here ?
I am the original author of the module.
I found a bug on it, and I sent a PR to fix it. 
I also added passing integration tests to it.

I implemented every suggestion give by @Lujeni @gundalow and @mattclay 

what else needs to be done for this to be merged?

marcosdiez commented:
@dagwieers I beleive I did everything that you requested. Could you review and approve my PR ? It's contains only bugfixes and examples, no new features.

marcosdiez commented:
@dagwieers please don't forget me!

marcosdiez commented:
@dagwieers , @gundalow I still don't know why this PR has not been merged...

marcosdiez commented:
@dagwieers , @gundalow I still don't know why this PR has not been merged...

marcosdiez commented:
@dagwieers , @gundalow I still don't know why this PR has not been merged...

marcosdiez commented:
@dagwieers , @gundalow I still don't know why this PR has not been merged...

gundalow commented:
bot_status

marcosdiez commented:
@rhysmeister I believe you are the person in charge for everything MongoDB related in Ansible.
I have this PR that has been opened since 2018 and all it needs is a code review and an approval.
Could you please take a look ? Thank you!

rhysmeister commented:
Hello @marcosdiez actually I'm not. I've just submitted a few PRs. I don't have any authority I'm afraid. 

There has been some chatter about moving the mongodb modules to a community support collection. Perhaps you'd like this included?

marcosdiez commented:
Hy Rhys!

Honestly, all I want is my 2 year old PR merged.

If you could code review it, then it would be easier to get the merge approved.

If not, well, please include it in the mongodb collection.

If you know anything I can do to get it approved please let me know.

Thanks

Marcos

Am 5. Februar 2020 21:40:28 MEZ schrieb Rhys Campbell <notifications@github.com>:
>Hello @marcosdiez actually I'm not. I've just submitted a few PRs. I
>don't have any authority I'm afraid. 
>
>There has been some chatter about moving the mongodb modules to a
>community support collection. Perhaps you'd like this included?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ansible/ansible/pull/46579#issuecomment-582601811

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

rhysmeister commented:
Hello Marcos,

I've tested your plugin... https://github.com/rhysmeister/TestMongoDBLookup

Looks good although a couple of comments. Error given when pymongo installed...

> fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'mongodb'. Error was a <class 'ansible.errors.AnsibleError'>, original message: pymongo is required in the master node (this machine) for mongodb lookup."}
 
Probably better without the "unhandled exception" stuff?

I also got the following exception...

> fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'mongodb'. Error was a <class 'KeyError'>, original message: 'database'"}

I realise that's because I was missing a few mandatory parameters but I think the message could be nicer/ better handled.

That's about all I can do at the moment. Good luck!

Rhys

marcosdiez commented:
Hey, @rhysmeister ,
thanks for testing my PR.

But following your instructions, you did not test the code from the PR, just the normal plugin in ansible.

I believe this is how it should be tested:

```
docker run -tid --rm -p 27017:27017 --name mongodb rhyscampbell/dockermongodb

virtualenv venv # or python3 -m venv venv
source venv/bin/activate
git clone --depth 1 --single-branch --branch fix_mongodb_lookup git@github.com:marcosdiez/ansible.git
source ansible/hacking/env-setup
pip install jinja2 
pip install pyyaml
ansible-playbook -l localhost test_mongodb.yml
```

the result is:

fatal: [localhost]: FAILED! => {"msg": "An unhandled exception occurred while running the lookup plugin 'mongodb'. Error was a <class 'ansible.errors.AnsibleError'>, original message: pymongo is required in the control node (this machine) for mongodb lookup."}

after `pip instsall pymongo` ansible runs successfully.

rhysmeister commented:
Hello @marcosdiez 

Your plugin is included in the lookup_plugins directory. Ansible will pick this up rather than the default (old) one. You can see it does because it creates a python cache dir...

ls -lh lookup_plugins/
total 24
drwxr-xr-x  3 rhyscampbell  staff    96B 14 Feb 12:02 __pycache__
-rw-r--r--  1 rhyscampbell  staff   9,6K 14 Feb 12:03 mongodb.py

You can add a syntax error to the file if you're not convinced. The playbook then breaks, i.e.

> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: NameError: name 'sdfsdfgsdagfsdfg' is not defined
> fatal: [localhost]: FAILED! => {"msg": "Unexpected failure during module execution."}

Rhys

marcosdiez commented:
Sorry, @rhysmeister, you are right.
I tried again and this time it happend exactly as you mentioned.

marcosdiez commented:
Hey @klingac

I have  PR to fix issues in Ansible's MongoDB lookup that is opened for more than 3 years. All I need is somebody like you to approve it so it can be merged.

Could you please take a look at it?

Thank you
Marcos

klingac commented on test/integration/targets/lookup_mongodb/tasks/main.yml:
```diff
@@ -0,0 +1,36 @@
+
+
+- include: test_lookup_mongo.yml
+  when: ansible_distribution == 'Ubuntu' and ansible_distribution_release == "xenial"
```
Is check for Xenial really necessery? We have at least newer LTS version for Ubuntu. I just tried that with 18.04 and it works.
```suggestion
  when: ansible_distribution == 'Ubuntu'
```

mdiez-modus replied:
No. It's not and it has been removed. When this was done, bionic (18.04) did not exist, xenial (16.04) was new and there were still ansible tests for 14.04, which either did not contain mongodb or pymongo, which is why this lines was there. Thanks for noticing.

marcosdiez replied:
No. It's not and it has been removed. When this was done, bionic (18.04) did not exist, xenial (16.04) was new and there were still ansible tests for 14.04, which either did not contain mongodb or pymongo, which is why this lines was there. Thanks for noticing.

klingac commented on lib/ansible/plugins/lookup/mongodb.py:
```diff
@@ -62,7 +62,9 @@
+                - Please use the strings C(ASCENDING) and C(DESCENDING) to set the order.
+                - Check the example for more information.
             type: list
             default: "[]"
```
Plase also add extra_parameters. You are mentioning it in examples, but not in documentation. It could be little bit confusing
```suggestion
            default: "[]"
        extra_connection_parameters:
            description:
                - Extra connection parameters, eg. { "readPreference": "nearest", "ssl" : True , "ssl_certfile": "/etc/self_signed_certificate.pem" }
            type: 'dict'
            default: '{}'
```

marcosdiez replied:
Once again, thanks for having noticed this. It's now added into the documentation as well.

klingac reviewed (commented):
Hello Marcos,

I've just tested your plugin on Ubuntu 18.04 LTS. Everything seems to be OK for me but I have two suggestions - see below. Those suggestion are only about tests and documentation. Plugin seems to be functioning well.

marcosdiez commented:
@klingac I did the two changes you requested and the tests are passing. Do you need anything ? If not, could you please approve this PR so it can get merged ? Thanks!

klingac reviewed (approved):
Looks OK to me.
Please merge this PR

marcosdiez commented:
@dagwieers I just got an approval from @klingac . Is this PR now eliable to be merged ? If not, what else is needed?

marcosdiez commented:
@dagwieers Please take a look at this PR!

gundalow commented:
rebuild_merge

gundalow commented:
FYI mongodb will soon have its own collection. For those wishing to join the mongodb working group you way wish to watch (subscribe) to https://github.com/ansible-collections/mongodb

The code that's currently in Ansible/Ansible will be moved to this new location within the next he next month.

See https://github.com/ansible-collections/overview for more info.

mattclay commented on test/integration/targets/lookup_mongodb/aliases:
```diff
@@ -0,0 +1,5 @@
+skip/freebsd
+skip/osx
+skip/rhel
+destructive
```
Add one more entry here:

```
skip/python2.6  # filters are controller only, and we no longer support Python 2.6 on the controller
```

mkrizek commented:
!waffling needs_revision

marcosdiez commented:
So... I pass all tests, what's next ?

marcosdiez commented:
Since this is a PR to fix a bug and I am trying to merge it for the last 2 years, could we just merge it against 2.9.x as a bugfix ?

sivel commented:
This PR would need to be resubmitted to https://github.com/ansible-collections/mongodb to be considered for merge.

All mongodb functionality has been removed from this repository and migrated to the above repo.

marcosdiez commented:
This PR has been moved here https://github.com/ansible-collections/mongodb/pull/38 and has been merged!
